worlds/wl4/__init__.py

```diff
@@ -0,0 +1,240 @@
+        cds = 16
+        full_health_items = (9, 7, 6)[difficulty.value]
+        treasures = 12 * treasure_hunt
+        total_required_locations = gem_pieces + cds + full_health_items + treasures
+
+        itempool = []
+
+        required_jewels = self.options.required_jewels.value
+        pool_jewels = self.options.pool_jewels.value
+        for name, item in filter_items(type=ItemType.JEWEL):
+            force_non_progression = required_jewels == 0
+            if item.passage() == Passage.ENTRY:
+                copies = min(pool_jewels, 1)
+            elif item.passage() == Passage.GOLDEN:
+                copies = self.options.golden_jewels.value
+                if self.options.goal.is_treasure_hunt():
+                    force_non_progression = True
+            else:
+                copies = pool_jewels
+
+            for _ in range(copies):
+                itempool.append(self.create_item(name, force_non_progression))
+
+        for name in filter_item_names(type=ItemType.CD):
+            itempool.append(self.create_item(name))
+
+        for name in filter_item_names(type=ItemType.ABILITY):
+            itempool.append(self.create_item(name))
+            if name.startswith('Progressive'):
+                itempool.append(self.create_item(name))
+
+        # Remove full health items to make space for abilities
+        if required_jewels == 4:
+            if difficulty == 0:
+                full_health_items -= 8
+            else:
+                raise ValueError('Not enough locations to place abilities for '
+                                 f'{self.multiworld.player_name[self.player]}. '
+                                 'Set the "Required Jewels" setting to a lower '
+                                 'value and try again.')
+
+        for _ in range(full_health_items):
+            itempool.append(self.create_item('Full Health Item'))
+
+        if (treasure_hunt):
+            for name in filter_item_names(type=ItemType.TREASURE):
+                itempool.append(self.create_item(name))
+
+        junk_count = total_required_locations - len(itempool)
+        junk_item_pool = tuple(filter_item_names(type=ItemType.ITEM))
+        for _ in range(junk_count):
+            item_name = self.multiworld.random.choice(junk_item_pool)
+            itempool.append(self.create_item(item_name))
+
+        self.multiworld.itempool += itempool
+
+    def generate_output(self, output_directory: str):
+        output_path = Path(output_directory)
+
+        patch = WL4ProcedurePatch()
+        patch.write_file('basepatch.bsdiff', data_path('basepatch.bsdiff'))
+        write_tokens(self, patch)
+        patch.procedure.append((
+            'shuffle_music_and_wario_voice',
+            [self.options.music_shuffle.value, self.options.wario_voice_shuffle.value]
+        ))
+
+        output_filename = self.multiworld.get_out_file_name_base(self.player)
+        patch.write(output_path / f'{output_filename}{patch.patch_file_ending}')
```

Comment:
I don't really know anything about this, but mypy gets mad about this being a `Path` and not a `str` so is the whole things supposed to be in the f-string?